File: backend/app/cards/validation.py
"""Generic input validation shared by every card, driven entirely by FieldSpec.

Lives once, here, so a new card never needs to write its own bounds/type checks.
"""

from typing import Any

from .schema import FieldKind, FieldSpec


class InputError(ValueError):
    """A validation failure carrying a machine-readable code beside its message.

    The code exists for the TypeScript parity corpus. Asserting that both
    implementations produce the same *message* is a trap: these embed ``{value!r}``,
    and Python renders ``True``/``'x'`` where JavaScript renders ``true``/``"x"`` --
    so message equality would mean hand-writing a Python ``repr`` in TypeScript
    forever, and would break on values nobody thought about.

    The message itself is unchanged, so the API's 422 body and every existing test
    still see exactly what they did before.
    """

    def __init__(self, field: str, code: str, message: str) -> None:
        super().__init__(message)
        self.field = field
        self.code = code


def validate_inputs(fields: list[FieldSpec], raw_inputs: dict[str, Any]) -> dict[str, Any]:
    """Apply each field's default for omitted keys, then type/bounds-check every value."""
    return {
        field.name: _validate_field(field, raw_inputs.get(field.name, field.default))
        for field in fields
    }


def _validate_field(field: FieldSpec, value: Any) -> Any:
    match field.kind:
        case FieldKind.BOOLEAN:
            return _validate_boolean(field, value)
        case FieldKind.NUMBER | FieldKind.COUNTER:
            return _validate_number(field, value)
        case FieldKind.SELECT:
            return _validate_select(field, value)
        case FieldKind.SEQUENCE:  # pragma: no branch -- exhaustive over FieldKind's 5 members
            return _validate_sequence(field, value)


def _validate_boolean(field: FieldSpec, value: Any) -> bool:
    if not isinstance(value, bool):
        raise InputError(
            field.name, "not_boolean", f"{field.name!r} must be a boolean, got {value!r}"
        )
    return value


def _validate_number(field: FieldSpec, value: Any) -> int:
    # bool is a subclass of int in Python, so True would otherwise pass as 1. No such
    # hazard exists in the TypeScript port, which is exactly why the corpus compares
    # error *codes* rather than trying to mirror this check.
    if isinstance(value, bool) or not isinstance(value, int):
        raise InputError(
            field.name, "not_integer", f"{field.name!r} must be an integer, got {value!r}"
        )
    if field.min is not None and value < field.min:
        raise InputError(
            field.name, "below_min", f"{field.name!r} must be >= {field.min}, got {value}"
        )
    if field.max is not None and value > field.max:
        raise InputError(
            field.name, "above_max", f"{field.name!r} must be <= {field.max}, got {value}"
        )
    return value


def _validate_select(field: FieldSpec, value: Any) -> str:
    if not isinstance(value, str):
        raise InputError(
            field.name, "not_string", f"{field.name!r} must be a string, got {value!r}"
        )
    allowed = {option.value for option in field.options or []}
    if value not in allowed:
        raise InputError(
            field.name,
            "not_in_options",
            f"{field.name!r} must be one of {sorted(allowed)}, got {value!r}",
        )
    return value


def _validate_sequence(field: FieldSpec, value: Any) -> list[str]:
    """Order-preserving list of `options` values, capped at `max` entries.

    Returns a *copy*: a sequence field's default is a shared list living on the
    module-level METADATA singleton, so handing it straight back would let one
    request's inputs alias every later request's default.
    """
    if not isinstance(value, list):
        raise InputError(field.name, "not_list", f"{field.name!r} must be a list, got {value!r}")
    if field.max is not None and len(value) > field.max:
        raise InputError(
            field.name,
            "too_long",
            f"{field.name!r} must have at most {field.max} entries, got {len(value)}",
        )
    allowed = {option.value for option in field.options or []}
    for entry in value:
        if entry not in allowed:
            raise InputError(
                field.name,
                "entry_not_in_options",
                f"{field.name!r} entries must each be one of {sorted(allowed)}, got {entry!r}",
            )
    if field.map is not None:
        _validate_walk(field, value)
    return list(value)


def _validate_walk(field: FieldSpec, value: list[str]) -> None:
    """A mapped sequence must be a legal walk: start at the entry, then follow an
    edge for every step. Venture never moves backwards or teleports (CR 309), and the
    UI only offers legal successors -- so this rejects only a hand-crafted request,
    the same defensive posture the membership check above already takes."""
    assert field.map is not None  # narrowed by the caller; mypy cannot see it
    if not value:
        return
    if value[0] != field.map.entry:
        raise InputError(
            field.name,
            "illegal_room",
            f"{field.name!r} must start at {field.map.entry!r}, got {value[0]!r}",
        )
    successors: dict[str, set[str]] = {}
    for edge in field.map.edges:
        successors.setdefault(edge.source, set()).add(edge.target)
    for previous, current in zip(value, value[1:], strict=False):
        if current not in successors.get(previous, set()):
            raise InputError(
                field.name,
                "illegal_room",
                f"{field.name!r} cannot move from {previous!r} to {current!r}",
            )

File: packages/core/src/cards/validation.ts
import type { FieldSpec, FieldValues, MapSpec } from '../types'

/**
 * The TypeScript half of card input validation. Mirrors `backend/app/cards/validation.py`
 * field-for-field; the parity corpus asserts they agree.
 *
 * Errors carry a `code` rather than only a message. Python's messages embed `repr()`,
 * which renders `True`/`'x'` where JavaScript renders `true`/`"x"` -- so asserting
 * message equality across the two would mean maintaining a Python `repr` in TypeScript
 * forever, and would break on values nobody anticipated. The code is a small closed set
 * both sides agree on; the message is each language's own.
 */
export type InputErrorCode =
  | 'not_boolean'
  | 'not_integer'
  | 'below_min'
  | 'above_max'
  | 'not_string'
  | 'not_in_options'
  | 'not_list'
  | 'too_long'
  | 'entry_not_in_options'
  | 'illegal_room'

export class InputError extends Error {
  // Declared explicitly rather than as constructor parameter properties, which are a
  // TypeScript-only construct and so barred by `erasableSyntaxOnly`.
  readonly field: string
  readonly code: InputErrorCode

  constructor(field: string, code: InputErrorCode, message: string) {
    super(message)
    this.name = 'InputError'
    this.field = field
    this.code = code
  }
}

function allowedValues(field: FieldSpec): string[] {
  return (field.options ?? []).map((option) => option.value)
}

function validateBoolean(field: FieldSpec, value: unknown): boolean {
  if (typeof value !== 'boolean') {
    throw new InputError(field.name, 'not_boolean', `${field.name} must be a boolean`)
  }
  return value
}

function validateNumber(field: FieldSpec, value: unknown): number {
  // Python additionally rejects `True`, since bool subclasses int there. No such
  // hazard here -- a boolean is not a number in JavaScript -- which is exactly why
  // the corpus compares codes rather than trying to mirror the check itself.
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    throw new InputError(field.name, 'not_integer', `${field.name} must be an integer`)
  }
  if (field.min != null && value < field.min) {
    throw new InputError(field.name, 'below_min', `${field.name} must be >= ${field.min}`)
  }
  if (field.max != null && value > field.max) {
    throw new InputError(field.name, 'above_max', `${field.name} must be <= ${field.max}`)
  }
  return value
}

function validateSelect(field: FieldSpec, value: unknown): string {
  if (typeof value !== 'string') {
    throw new InputError(field.name, 'not_string', `${field.name} must be a string`)
  }
  const allowed = allowedValues(field)
  if (!allowed.includes(value)) {
    throw new InputError(
      field.name,
      'not_in_options',
      `${field.name} must be one of ${[...allowed].sort().join(', ')}`,
    )
  }
  return value
}

function validateSequence(field: FieldSpec, value: unknown): string[] {
  if (!Array.isArray(value)) {
    throw new InputError(field.name, 'not_list', `${field.name} must be a list`)
  }
  if (field.max != null && value.length > field.max) {
    throw new InputError(
      field.name,
      'too_long',
      `${field.name} must have at most ${field.max} entries`,
    )
  }
  const allowed = allowedValues(field)
  for (const entry of value) {
    if (typeof entry !== 'string' || !allowed.includes(entry)) {
      throw new InputError(
        field.name,
        'entry_not_in_options',
        `${field.name} entries must each be one of ${[...allowed].sort().join(', ')}`,
      )
    }
  }
  if (field.map != null) {
    validateWalk(field, field.map, value as string[])
  }
  // A copy, for the same reason Python returns one: a sequence field's default is a
  // shared array on the metadata singleton, so handing it back would let one session's
  // inputs alias every later default.
  return [...value]
}

/** A mapped sequence must be a legal walk: start at the entry, then follow an edge for
 * every step. Venture never moves backwards or teleports (CR 309); the UI only offers
 * legal successors, so this rejects only a hand-crafted input -- the same defensive
 * posture the membership check above already takes. */
function validateWalk(field: FieldSpec, map: MapSpec, value: string[]): void {
  if (value.length === 0) return
  if (value[0] !== map.entry) {
    throw new InputError(
      field.name,
      'illegal_room',
      `${field.name} must start at ${map.entry}`,
    )
  }
  const successors = new Map<string, Set<string>>()
  for (const edge of map.edges) {
    let targets = successors.get(edge.source)
    if (!targets) {
      targets = new Set()
      successors.set(edge.source, targets)
    }
    targets.add(edge.target)
  }
  for (let i = 1; i < value.length; i += 1) {
    if (!successors.get(value[i - 1])?.has(value[i])) {
      throw new InputError(
        field.name,
        'illegal_room',
        `${field.name} cannot move from ${value[i - 1]} to ${value[i]}`,
      )
    }
  }
}

/** Applies each field's default for omitted keys, then type/bounds-checks every value. */
export function validateInputs(fields: FieldSpec[], rawInputs: FieldValues): FieldValues {
  const validated: FieldValues = {}
  for (const field of fields) {
    const value = field.name in rawInputs ? rawInputs[field.name] : field.default
    switch (field.kind) {
      case 'boolean':
        validated[field.name] = validateBoolean(field, value)
        break
      case 'number':
      case 'counter':
        validated[field.name] = validateNumber(field, value)
        break
      case 'select':
        validated[field.name] = validateSelect(field, value)
        break
      case 'sequence':
        validated[field.name] = validateSequence(field, value)
        break
    }
  }
  return validated
}
